### PR TITLE
Allow users to access VariantDebugInfo by setting a flag (E2-1283)

### PR DESCRIPTION
**Pull request by e7mac**
PRD: https://leanplum.atlassian.net/wiki/spaces/PRODDEV/pages/548503556/Tinder+Expose+granular+information+on+variables+and+A+B+tests+to+SDK

We allow a flag to be set in "includeVariantDebugInfo" that retrieves content assignment details from the server. We then allow access to this object through Leanplum.variantDebugInfo

### CODE REVIEW

**Comment by e7mac** [comment]
Test this please

**Comment by e7mac** [comment]
test this please

File: Example/Tests/Classes/LeanplumTest.m [hunk 1]
```diff
@@ -208,6 +208,50 @@ - (void) test_complex_development_start
...
+    [Leanplum setVariantDebugInfoEnabled:YES];
+    [OHHTTPStubs stubRequestsPassingTest:^BOOL(NSURLRequest * _Nonnull request) {
+        return [request.URL.host isEqualToString:API_HOST];
+    } withStubResponse:^OHHTTPStubsResponse * _Nonnull(NSURLRequest * _Nonnull request) {
+        NSString *response_file = OHPathForFile(@"simple_start_response.json", self.class);
+        return [OHHTTPStubsResponse responseWithFileAtPath:response_file statusCode:200
+                                                   headers:@{@"Content-Type":@"application/json"}];
+    }];
+    
+    // Validate requst.
```
Comment (rossenassenov) [inline Example/Tests/Classes/LeanplumTest.m]: typo: request

File: Example/Tests/Classes/LeanplumTest.m [hunk 2]
```diff
@@ -1122,6 +1166,70 @@ - (void)test_variables
     XCTAssertEqualObjects(@100LL, [long_long_variable defaultValue]);
 }
 
+/**
+ * Tests variant debug info.
+ */
+- (void)testStartResponseShouldParseVariantDebugInfo
+{
+    [Leanplum setVariantDebugInfoEnabled:YES];
```
Comment (rossenassenov) [inline Example/Tests/Classes/LeanplumTest.m]: test style: adding a few comments for `given/when/then` will make this test much easier to understand and maintain in the future

File: Example/Tests/Classes/LeanplumTest.m [hunk 3]
```diff
@@ -1122,6 +1166,70 @@ - (void)test_variables
...
+        XCTAssertNotNil([LPVarCache variantDebugInfo]);
+        dispatch_semaphore_signal(semaphore);
+    }];
+    dispatch_semaphore_wait(semaphore, [LeanplumHelper default_dispatch_time]);
+    
+    // Remove stub after start is successful, so we don't capture requests from other methods
+    [OHHTTPStubs stubRequestsPassingTest:^BOOL(NSURLRequest * _Nonnull request) {
+        return [request.URL.host isEqualToString:API_HOST];
+    } withStubResponse:^OHHTTPStubsResponse * _Nonnull(NSURLRequest * _Nonnull request) {
+        NSString* response_file = OHPathForFile(@"variables_response.json", self.class);
```
Comment (rossenassenov) [inline Example/Tests/Classes/LeanplumTest.m]: should we always be returning variables_response.json in other tests ?
is any other test doing this stubbing ?

File: Example/Tests/Classes/LeanplumTest.m [hunk 4]
```diff
@@ -1122,6 +1166,70 @@ - (void)test_variables
...
+                                                   headers:@{@"Content-Type":@"application/json"}];
+    }];
+}
+
+/**
+ * Tests variant debug info persistence.
+ */
+- (void)testShouldPersistVariantDebugInfo
+{
+    NSDictionary *mockVariantDebugInfo = @{@"abTests":@[]};
```
Comment (rossenassenov) [inline Example/Tests/Classes/LeanplumTest.m]: move this mockVariantDebugInfo assignment before line 1220 where it's actually used

Comment (rossenassenov) [inline Example/Tests/Classes/LeanplumTest.m]: add a few comments to document and explain the expected behavior

File: Example/Tests/Resources/Responses/start_with_variant_debug_info_response.json [hunk 5]
```diff
@@ -0,0 +1,296 @@
+{
```
Comment (rossenassenov) [inline Example/Tests/Resources/Responses/start_with_variant_debug_info_response.json]: autoformat this json file

**Comment by rossenassenov** [comment]
please put the jira ticket number in the title of the PR, makes it easier to understand git history

File: Example/Tests/Resources/Responses/start_with_variant_debug_info_response.json (hunk 5 above)
Comment (e7mac) [reply Example/Tests/Resources/Responses/start_with_variant_debug_info_response.json]: sadly, this is the automatic xcode format. I'll do it through another IDE for this

**Pull request merged**